server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java

```diff
@@ -313,65 +351,72 @@ private static IndexRequest createSynonymRuleIndexRequest(String synonymsSetId,
-            .setSize(1)
-            .setPreference(Preference.LOCAL.type())
+        // Search for the synonym set ID
+        client.prepareGet(SYNONYMS_ALIAS_NAME, synonymsSetId)
             .execute(new DelegatingIndexNotFoundActionListener<>(synonymsSetId, listener, (l, searchResponse) -> {
-                if (searchResponse.getHits().getTotalHits().value == 0) {
+                if (searchResponse.isExists() == false) {
                     l.onFailure(new ResourceNotFoundException("Synonym set [" + synonymsSetId + "] not found"));
                     return;
                 }
                 l.onResponse(null);
             }));
     }
 
-    // Deletes a synonym set rules, using the supplied listener
-    private void deleteSynonymsSetRules(String resourceName, ActionListener<BulkByScrollResponse> listener) {
-        // Delete synonyms set if it existed previously. Avoid catching an index not found error by ignoring unavailable indices
+    // Deletes a synonym set rules and the synonym set object, using the supplied listener
+    private void deleteSynonymsSetObjects(String synonymSetId, ActionListener<BulkByScrollResponse> listener) {
         DeleteByQueryRequest dbqRequest = new DeleteByQueryRequest(SYNONYMS_ALIAS_NAME).setQuery(
-            QueryBuilders.termQuery(SYNONYMS_SET_FIELD, resourceName)
+            QueryBuilders.termQuery(SYNONYMS_SET_FIELD, synonymSetId)
         ).setRefresh(true).setIndicesOptions(IndicesOptions.fromOptions(true, true, false, false));
 
         client.execute(DeleteByQueryAction.INSTANCE, dbqRequest, listener);
     }
 
-    public void deleteSynonymsSet(String resourceName, ActionListener<SynonymsReloadResult<AcknowledgedResponse>> listener) {
-        deleteSynonymsSetRules(resourceName, listener.delegateFailure((l, bulkByScrollResponse) -> {
-            if (bulkByScrollResponse.getDeleted() == 0) {
-                // If nothing was deleted, synonym set did not exist
-                l.onFailure(new ResourceNotFoundException("Synonym set [" + resourceName + "] not found"));
-                return;
-            }
-            final List<BulkItemResponse.Failure> bulkFailures = bulkByScrollResponse.getBulkFailures();
-            if (bulkFailures.isEmpty() == false) {
-                l.onFailure(
-                    new ElasticsearchException(
-                        "Error deleting synonym set: "
-                            + bulkFailures.stream().map(BulkItemResponse.Failure::getMessage).collect(Collectors.joining("\n"))
-                    )
-                );
-                return;
-            }
-
-            reloadAnalyzers(resourceName, l, AcknowledgedResponse.of(true));
+    public void deleteSynonymsSet(String synonymSetId, ActionListener<SynonymsReloadResult<AcknowledgedResponse>> listener) {
+        // Check it exists to simplify code paths
+        checkSynonymSetExists(synonymSetId, listener.delegateFailure((checkListener, response) -> {
```

Comment:
Can you please explain why do we need this extra check here, why we can't just delete?

Reply:
We must check whether the synonym set actually exist to return a 404 otherwise.

Reply:
Can we check `bulkByScrollResponse.getDeleted() == 0`  it was before. and if it is equal to 0, throw ResourceNotFoundException as it was done before? WDYT?

Reply:
yes, we can do that. Changed